### PR TITLE
Join implicit concatenated strings when they fit on a line

### PR DESCRIPTION
**Pull request by MichaReiser**
## Summary

Implements https://github.com/astral-sh/ruff/issues/9457 for f-strings, regular strings, and byte literals. The implementation doesn't handle raw strings or triple quoted strings, similar to black's string-processing preview style.

Here an example from the ecosystem check. The formatter now joins the implicit concatenated f-string into a single string literal.

```diff
             raise TypeError(
-                "Expected str or callable for parameter 'probe', "
-                f"not '{method.__class__.__name__}'"
+                f"Expected str or callable for parameter 'probe', not '{method.__class__.__name__}'"
             )
```

What makes this change "complicated" is that Ruff has already many special casing for strings:

* Comments at the end of an assignment where the value is a string gets inlined:
  ```python
  aaaaaaaaaaaaa = (
       "long string literal"  # comment
  )

  # instead  of 
  aaaaaaaaaaaaa = (
       "long string literal"
  )  # comment
  ```
* Docstring formatting trims leading and trailing whitespace

In addition, there are many other string-related preview styles, and this implementation must be compatible with the current stable formatting and any other combination of preview styles. Specifically, the implementation must support the stable and preview f-string formatting.

## `assert` formatting

This new style requires a few smaller changes in how expressions containing strings are formatted to mitigate instabilities. However, there's one large change that I'm undecided if we should move forward with it. 

The existing assertion formatting is prone to unstable formatting when the right is an implicit concatenated string that can be joined. The easiest fix was to change the formatting to use `Parenthesize::IfBreaksParenthesized` which we use in other places where we format multiple expressions with the `maybe_parenthesize` layout. However, it has the effect that assertions now break from right to left: We always parenthesize the right before we start splitting the left. This impacts formatting of almost all `assert` statements that are longer than the configured line length and the right is a string because we used to split the assertion and we now parenthesize the message instead. 

Here two examples from the ecosystem checks:

```diff
     with conn.read_ctx() as cursor:  # make sure it wrote in the DB but not the last one
-        assert cursor.execute("SELECT b from a").fetchall() == [
-            (1,),
-            (2,),
-            (3,),
-            (4,),
-        ], "other greenlet should write to the DB"  # noqa: E501
+        assert cursor.execute("SELECT b from a").fetchall() == [(1,), (2,), (3,), (4,)], (
+            "other greenlet should write to the DB"
+        )  # noqa: E501
```

```diff
                         new_id = f"{id}{suffix}{id_suffixes[id]}"
                     resolved_ids[index] = new_id
                     id_suffixes[id] += 1
-        assert len(resolved_ids) == len(
-            set(resolved_ids)
-        ), f"Internal error: {resolved_ids=}"
+        assert len(resolved_ids) == len(set(resolved_ids)), (
+            f"Internal error: {resolved_ids=}"
+        )
         return resolved_ids
 
     def _resolve_ids(self) -> Iterable[str]:
```

You can see how the formatter now keeps the assertion flat (left) and parenthesizes the message. I do find that this overall improves the readability and is also in line with what we do in assignments: We parenthesize the value to avoid splitting the target. 

I have three concerns with the style change:

**1. Black compatibility/Upgrade diff**

IMO, this is the most important one. Our assert formatting has never been compatible with black and this has come up in the past (https://github.com/astral-sh/ruff/issues/13386, https://github.com/astral-sh/ruff/issues/8331, https://github.com/astral-sh/ruff/issues/8388), but there's no clear consensus from users on which style they prefer. This PR would deviate our implementation further from Black's style and it will lead to larger diffs when migrating from Black but it also leads to large diffs when upgrading to the new style.

**2. More parentheses**

Black overall tries to reduce the parentheses. For example, Black prefers

```python
if aaaaaaaaaaaaaa + bbbbbbbbbbbbbbb / len([
    122222,
    33333333333333,
    44444444444444,
    55444,
]):
    ...
```

over 

```python
if (
    aaaaaaaaaaaaaa
    + bbbbbbbbbbbbbbb
    / len([122222, 33333333333333, 44444444444444, 55444])
):
    ...
```

when the left-most or right-most expression has its own parentheses (and some more rules). Black also applies this to assert formatting (`-: Black, +: Ruff`). 

```diff

-            assert isinstance(
-                document, dict
-            ), "document should be of type `dict[str,Any]`. But found: `{}`".format(
-                type(document)
+            assert isinstance(document, dict), (
+                "document should be of type `dict[str,Any]`. But found: `{}`".format(
+                    type(document)
+                )
             )
```

```diff
-        assert isinstance(
-            input, dict
-        ), "The input to RunnablePassthrough.assign() must be a dict."
+        assert isinstance(input, dict), (
+            "The input to RunnablePassthrough.assign() must be a dict."
+        )
 
```

You can see how Black uses the parentheses of the call expression over introducing new parentheses to break the right. 

IMO, this makes assert statements harder to read because it is unclear where the assertion ends and the message start. However, it helps to reduce the overall number of parentheses

**3. assertions that end in parentheses**

There's one cases where the new formatting is arguably worse:

```python
# Black

assert graph.artifacts == [
    GraphArtifact(
        id=expected_top_level_artifact.id,
        created=expected_top_level_artifact.created,
        key=expected_top_level_artifact.key,
        type=expected_top_level_artifact.type,
        data=(
            expected_top_level_artifact.data
            if expected_top_level_artifact.type == "progress"
            else None
        ),
        is_latest=True,
    )
], "Expected artifacts associated with the flow run but not with a task to be included at the roof of the graph."

# Ruff 
assert graph.artifacts == [
    GraphArtifact(
        id=expected_top_level_artifact.id,
        created=expected_top_level_artifact.created,
        key=expected_top_level_artifact.key,
        type=expected_top_level_artifact.type,
        data=expected_top_level_artifact.data
        if expected_top_level_artifact.type == "progress"
        else None,
        is_latest=True,
    )
], (
    "Expected artifacts associated with the flow run but not with a task to be included at the roof of the graph."
)
```

In this example, parenthesizing the message is *useless* because it doesn't help to get more space for the message. The indent requires as much space as the `], (`.

 

The reason I went with this change (and am seeking feedback now) is that:

* It's an extremely easy fix for the instability
* We use break right over left in other places. Therefore, it's arguably more consistent for Ruff
* I do find that it improves formatting in many cases

I do think it would be possible to implement something closer to today's style by using `best_fit` similar to the assignment formatting. And it might not even be that much work (a day or two). So I'm up to tackle this if we conclude that we want to keep closer to Black's style, reduce the upgrade diff, or are concerned about the "bad" cases.

## TODOs

* [x] Review for instabilities when an implicit concatenated string is the right side of an assignment where the target or the annotation is splittable. 
* [x] Review for instabilities with the hug expressions style
* [x] Review implicit concatenated string formatting in `maybe_parenthesize_expression` changes similar to assert stmt (e.g. `YieldExpr`)
* [x] Support the f-string formatting preview style (Big)
* [x] Test with `QuoteStyle` preserve
* [x] In doc-string positions
   ```python
    def in_docstring_position():
        "diffent '" 'quote "are fine"                '
    ```
* [x] Handle strings in `ExprStmt` positions (there's no group)
* [x] Review changes to larger repositories and the ecosystem results
* [x] ~~(stretch) Remove parentheses when collapsing f-strings https://play.ruff.rs/1355c0b0-f452-445a-8ffc-6285e1b0885e~~
* [x] Tests for `can_omit_optional_parentheses` changes.

## Review

* I would appreciate feedback on the assertion style changes.
* Look out if there are more string edge cases that need handling

@AlexWaygood I don't expect you to review the code changes but I highly value your input on the assertion formatting changes. 

## Test Plan

I added plenty of new integration tests and I reviewed the ecosystem changes and did test the formatting on some larger projects. 

Non-truncated ecosystem changes: https://gist.github.com/MichaReiser/c82d6132e577d9e7ee9ae85543904099

### CODE REVIEW

**Comment by MichaReiser** [comment]
I like what I'm seeing in the diff. I don't like seeing another instability, ugh

**Comment by MichaReiser** [comment]
This one seems unfortunate

```diff
         # Written on 06.06.2022 and can be removed after a couple of months if everything goes well
         assert (
             ZERO <= used_amount <= self._acquisitions_heap[0].acquisition_event.remaining_amount
-        ), f"Used amount must be in the interval [0, {self._acquisitions_heap[0].acquisition_event.remaining_amount}] but it was {used_amount} for {asset}"  # noqa: E501
+        ), (
+            f"Used amount must be in the interval [0, {self._acquisitions_heap[0].acquisition_event.remaining_amount}] but it was {used_amount} for {asset}"
+        )  # noqa: E501
 
         self._acquisitions_heap[0].acquisition_event.remaining_amount -= used_amount
         if self._acquisitions_heap[0].acquisition_event.remaining_amount == ZERO:
```

File: crates/ruff_python_formatter/src/string/implicit.rs [hunk 1]
```diff
@@ -0,0 +1,387 @@
...
+            // Only determining the preferred quote for the first string is sufficient
+            // because we don't support joining triple quoted strings with non triple quoted strings.
+            let quote = if let Ok(preferred_quote) =
+                Quote::try_from(normalizer.preferred_quote_style(first_part))
+            {
+                for part in string.parts() {
+                    let part_quote_metadata =
+                        QuoteMetadata::from_part(part, preferred_quote, context);
+
+                    if let Some(merged) = merged_quotes.as_mut() {
```
Comment (MichaReiser) [inline crates/ruff_python_formatter/src/string/implicit.rs:202]: TODO: This is not correct for f-strings

File: crates/ruff_python_formatter/src/string/docstring.rs [hunk 2]
```diff
@@ -1605,7 +1605,8 @@ fn docstring_format_source(
 /// that avoids `content""""` and `content\"""`. This does only applies to un-escaped backslashes,
 /// so `content\\ """` doesn't need a space while `content\\\ """` does.
 fn needs_chaperone_space(normalized: &NormalizedString, trim_end: &str) -> bool {
-    trim_end.ends_with(normalized.flags().quote_style().as_char())
+    normalized.flags().is_triple_quoted()
```
Comment (MichaReiser) [inline crates/ruff_python_formatter/src/string/docstring.rs:1608]: TODO: gate behind preview

File: crates/ruff_python_formatter/src/expression/mod.rs [hunk 3]
```diff
@@ -405,38 +406,39 @@ impl Format<PyFormatContext<'_>> for MaybeParenthesizeExpression<'_> {
...
                 Parenthesize::IfBreaksParenthesized | Parenthesize::IfBreaksParenthesizedNested => {
-                    parenthesize_if_expands(&expression.format().with_options(Parentheses::Never))
-                        .fmt(f)
+                    // Can-omit layout is relevant for `"abcd".call`. We don't want to add unnecessary
+                    // parentheses in this case.
+                    if can_omit_optional_parentheses(expression, f.context()) {
+                        optional_parentheses(&unparenthesized).fmt(f)
+                    } else {
+                        parenthesize_if_expands(&unparenthesized).fmt(f)
+                    }
```
Comment (MichaReiser) [inline crates/ruff_python_formatter/src/expression/mod.rs:438]: This is the main change. All other changes are replacing `expression.format().with_options(Parentheses::Never)` with `unparenthesized

**Comment by AlexWaygood** [comment]
> @AlexWaygood I don't expect you to review the code changes but I highly value your input on the assertion formatting changes.

Overall I find these changes _especially_ are a significant improvement to readability.

I tried running this branch on some repos I have clones for locally, and a _lot_ of the changes were related to this, but I found all of them to be improvements. I think there will undoubtedly be some specific patterns where formatting is likely to regress (such as `3. assertions that end in parentheses` in your PR description), but I'm quite strongly in favour of this change overall. I also think the formatting for that specific case would be quite easy for a user to fix manually by splitting the string over multiple lines (but I _don't_ think that's something that Ruff should try to do for the user):

```py
assert graph.artifacts == [
    GraphArtifact(
        id=expected_top_level_artifact.id,
        created=expected_top_level_artifact.created,
        key=expected_top_level_artifact.key,
        type=expected_top_level_artifact.type,
        data=expected_top_level_artifact.data
        if expected_top_level_artifact.type == "progress"
        else None,
        is_latest=True,
    )
], (
    "Expected artifacts associated with the flow run "
    "but not with a task to be included at the roof of the graph."
)
```

File: crates/ruff_python_formatter/src/statement/stmt_assign.rs [hunk 4]
```diff
@@ -301,28 +308,145 @@ impl Format<PyFormatContext<'_>> for FormatStatementsLastExpression<'_> {
...
+                    // ____aaa = (
+                    //     "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaabbbbbbbbbbbbbbbbbbbbbbbbbbbvvvvvvvvvvvvvvv"
+                    // )  # c
+                    // ```
+                    //
+                    // However, this now gets reformatted again because Ruff now takes the `BestFit` layout for the string
+                    // because the value is no longer an implicit concatenated string.
+                    // ```python
+                    // ____aaa = "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaabbbbbbbbbbbbbbbbbbbbbbbbbbbvvvvvvvvvvvvvvv"  # c
+                    // ```
```
Comment (charliermarsh) [inline crates/ruff_python_formatter/src/statement/stmt_assign.rs:333]: (Very helpful.)

**Review by charliermarsh — APPROVED** [review]
This is excellent. The code is so clear and easy to read (especially for the formatter). Well done with the comments, clarity, etc.

**Comment by charliermarsh** [comment]
I think the `assert` formatting is probably an improvement, though I hear your concerns. This one stuck out to me a bit:

```diff

-            assert isinstance(
-                document, dict
-            ), "document should be of type `dict[str,Any]`. But found: `{}`".format(
-                type(document)
+            assert isinstance(document, dict), (
+                "document should be of type `dict[str,Any]`. But found: `{}`".format(
+                    type(document)
+                )
             )
```

But it's actually _not_ an increase in total lines, and the condition and message are actually more-clearly separated than before.

File: crates/ruff_python_formatter/src/statement/stmt_assign.rs [hunk 5]
```diff
@@ -301,28 +308,145 @@ impl Format<PyFormatContext<'_>> for FormatStatementsLastExpression<'_> {
...
+                        //    2,
+                        // ]}" "more"
+                        // ```
+                        if string.is_fstring() && flat.inspect(f)?.will_break() {
+                            return maybe_parenthesize_expression(
+                                value,
+                                *statement,
+                                Parenthesize::IfBreaks,
+                            )
+                            .fmt(f);
```
Comment (dhruvmanila) [inline crates/ruff_python_formatter/src/statement/stmt_assign.rs:367]: Can you help me understand why shouldn't we concatenate these two strings?

**Review by dhruvmanila — APPROVED** [review]
This is really thorough! Thanks for taking the time to add comments throughout the code and especially in tests. It was really easy for me to look at the formatted output directly in test because the comment described the original source code and what to expect.

Regarding the assert formatting, I too think that it's improves readability as it's easier to split the condition and the message mentally.

File: crates/ruff_python_formatter/src/statement/stmt_assign.rs (hunk 5 above)
Comment (MichaReiser) [reply crates/ruff_python_formatter/src/statement/stmt_assign.rs:367]: The important bit here is the magic trailing comma at the end of the list-expression. It always makes the list expand over multiple line. That means, it's guaranteed that the f-string can never fit on a single line. That's why we don't want to apply the implicit concatenated layout.

Comment (dhruvmanila) [reply crates/ruff_python_formatter/src/statement/stmt_assign.rs]: Ah right, got it.

Comment (MichaReiser) [reply crates/ruff_python_formatter/src/statement/stmt_assign.rs]: But good call out. It helped me discover another bug :)

**Comment by MichaReiser** [comment]
Okay. Let's merge this as is and revisit the assert formatting decision depending on the preview style feedback

**Pull request merged**